### ISSUE TITLE
Cannot compile

### ISSUE THREAD

**Issue by Syntax3rror404** [ArmainAP/Unreal-Engine-Software-Occlusion-Culling#1]
```
0>------ Building 6 action(s) started ------
0>[1/6] Compile [x64] OcclusionPrimitiveContext.cpp
0>OcclusionPrimitiveContext.cpp(16): Error C2039 : 'ComponentId': is not a member of 'UStaticMeshComponent'
0>SceneInterface.h(59): Reference  : see declaration of 'UStaticMeshComponent'
0>[2/6] Compile [x64] OcclusionCullingSubsystem.cpp
0>OcclusionCullingSubsystem.cpp(158): Error C2039 : 'ComponentId': is not a member of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.generated.h(13): Reference  : see declaration of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.cpp(158): Error C2737 : 'FoundPrimitiveInfo': const object must be initialized
0>OcclusionCullingSubsystem.cpp(158): Error C3536 : 'FoundPrimitiveInfo': cannot be used before it is initialized
0>OcclusionCullingSubsystem.cpp(160): Error C2100 : illegal indirection
0>OcclusionCullingSubsystem.cpp(160): Error C2440 : 'initializing': cannot convert from 'int' to 'UOcclusionPrimitiveContext *'
0>OcclusionCullingSubsystem.cpp(160): Reference  : Conversion from integral type to pointer type requires reinterpret_cast, C-style cast or parenthesized function-style cast
0>OcclusionCullingSubsystem.cpp(167): Error C2039 : 'ComponentId': is not a member of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.generated.h(13): Reference  : see declaration of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.cpp(174): Error C2039 : 'ComponentId': is not a member of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.generated.h(13): Reference  : see declaration of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.cpp(192): Error C2039 : 'ComponentId': is not a member of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.generated.h(13): Reference  : see declaration of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.cpp(199): Error C2039 : 'ComponentId': is not a member of 'UStaticMeshComponent'
0>OcclusionCullingSubsystem.generated.h(13): Reference  : see declaration of 'UStaticMeshComponent'
0>[3/6] Compile [x64] Module.SoftwareOcclusionCulling.cpp
0>UniquePtr.h(64): Error C4150 : deletion of pointer to incomplete type 'FOcclusionFrameResults'; no destructor called
```

**Comment by ArmainAP** [comment]
Hello! Looks like `ComponentId` property is no longer available in latest engine version. 
I updated the code to use `GetPrimitiveSceneId()` in commit e86168e5fe280e7f67f0d5af6dff83bc4aeb1ade but I haven't fully tested it.

I will leave this issue open for a while in case you run into any other issues.

**Comment by Syntax3rror404** [comment]
> Hello! Looks like `ComponentId` property is no longer available in latest engine version. I updated the code to use `GetPrimitiveSceneId()` in commit [e86168e](https://github.com/ArmainAP/Unreal-Engine-Software-Occlusion-Culling/commit/e86168e5fe280e7f67f0d5af6dff83bc4aeb1ade) but I haven't fully tested it.
> 
> I will leave this issue open for a while in case you run into any other issues.

Thanks now the compiling works and the draw calls are insanly reduced!

Now we get this error while playing.

```
Unhandled Exception: EXCEPTION_ACCESS_VIOLATION reading address 0x00000000000000d0

UnrealEditor_SoftwareOcclusionCulling!UOcclusionPrimitiveContext::DebugBounds() [C:\Users\Documents\Unreal Projects\LuminaDev\Plugins\SoftwareOcclusionCulling\Source\SoftwareOcclusionCulling\Private\Data\OcclusionPrimitiveContext.cpp:139]
UnrealEditor_SoftwareOcclusionCulling!UOcclusionCullingSubsystem::PopulateScene() [C:\Users\Documents\Unreal Projects\LuminaDev\Plugins\SoftwareOcclusionCulling\Source\SoftwareOcclusionCulling\Private\OcclusionCullingSubsystem.cpp:219]
UnrealEditor_SoftwareOcclusionCulling!UOcclusionCullingSubsystem::Tick() [C:\Users\Documents\Unreal Projects\LuminaDev\Plugins\SoftwareOcclusionCulling\Source\SoftwareOcclusionCulling\Private\OcclusionCullingSubsystem.cpp:78]
```

And this 

```
Unhandled Exception: EXCEPTION_ACCESS_VIOLATION 0x000006288a3affc0
UnrealEditor_SoftwareOcclusionCulling!TArray<FOcclusionMeshData,TSizedDefaultAllocator<32> >::~TArray<FOcclusionMeshData,TSizedDefaultAllocator<32> >() [C:\Program Files\Epic Games\UE_5.4\Engine\Source\Runtime\Core\Public\Containers\Array.h:679]
UnrealEditor_SoftwareOcclusionCulling!UE::Core::Private::Function::TFunctionRefCaller<`SubmitScene'::`2'::<lambda_2>,void __cdecl(void)>::Call() [C:\Program Files\Epic Games\UE_5.4\Engine\Source\Runtime\Core\Public\Templates\Function.h:405]
UnrealEditor_SoftwareOcclusionCulling!TGraphTask<TFunctionGraphTaskImpl<void __cdecl(void),0> >::ExecuteTask() [C:\Program Files\Epic Games\UE_5.4\Engine\Source\Runtime\Core\Public\Async\TaskGraphInterfaces.h:1235]
```

Do I need to disable the default GPU occlusion first?

It crashes allways after the same amount of time.

While packaging there are also two errors:

```
UATHelper: Packaging (Android (ASTC)): Using Parallel executor to run 7 action(s)
UATHelper: Packaging (Android (ASTC)): ------ Building 7 action(s) started ------
UATHelper: Packaging (Android (ASTC)): [1/7] Compile [arm64] SoftwareOcclusionCulling.cpp
UATHelper: Packaging (Android (ASTC)): C:/Users/Documents/Unreal Projects/LuminaDev/Plugins/SoftwareOcclusionCulling/Source/SoftwareOcclusionCulling/Private/SoftwareOcclusionCulling.cpp(5,10): fatal error: 'ISettingsModule.h' file not found
UATHelper: Packaging (Android (ASTC)): #include "ISettingsModule.h"
UATHelper: Packaging (Android (ASTC)):          ^~~~~~~~~~~~~~~~~~~
UATHelper: Packaging (Android (ASTC)): 1 error generated.
PackagingResults: Error: 'ISettingsModule.h' file not found
UATHelper: Packaging (Android (ASTC)): [2/7] Compile [arm64] OcclusionCullingSubsystem.cpp
UATHelper: Packaging (Android (ASTC)): In file included from C:/Users/Documents/Unreal Projects/LuminaDev/Plugins/SoftwareOcclusionCulling/Source/SoftwareOcclusionCulling/Private/OcclusionCullingSubsystem.cpp:6:
UATHelper: Packaging (Android (ASTC)): C:/Users/Documents/Unreal Projects/LuminaDev/Plugins/SoftwareOcclusionCulling/Source/SoftwareOcclusionCulling/Private/Data/OcclusionViewInfo.h(6,10): fatal error: 'IXRTrackingSystem.h' file not found
UATHelper: Packaging (Android (ASTC)): #include "IXRTrackingSystem.h"
UATHelper: Packaging (Android (ASTC)):          ^~~~~~~~~~~~~~~~~~~~~
UATHelper: Packaging (Android (ASTC)): 1 error generated.
PackagingResults: Error: 'IXRTrackingSystem.h' file not found
UATHelper: Packaging (Android (ASTC)): [3/7] Compile [arm64] SoftwareOcclusionCullingOverride.cpp
UATHelper: Packaging (Android (ASTC)): [4/7] Compile [arm64] Module.SoftwareOcclusionCulling.cpp
UATHelper: Packaging (Android (ASTC)): [5/7] Compile [arm64] OcclusionPrimitiveContext.cpp
```

**Comment by ArmainAP** [comment]
Hello! Sorry about the lack of answer. It has been a busy week. 
I will take a look over the compile and crash issues.

> Do I need to disable the default GPU occlusion first?

They shouldn't collide, so it is up to you. Originally I made this because GPU occlusion culling was performing worse on mobile compared to CPU occlusion culling.

**Comment by Syntax3rror404** [comment]
> Hello! Sorry about the lack of answer. It has been a busy week. I will take a look over the compile and crash issues.
> 
> > Do I need to disable the default GPU occlusion first?
> 
> They shouldn't collide, so it is up to you. Originally I made this because GPU occlusion culling was performing worse on mobile compared to CPU occlusion culling.

Yes, my game only runs without occlusion culling on quest 2 with occlusion culling it is unplayable. On PC with a 4090 you get with occlusion culling 20ms more gpu time and the CPU cores are all emty :D HZB is better but costs also 15ms on the GPU.

So your plugin basicly fixes the engine, GPU occlusion culling is a bad idea.

**Comment by ArmainAP** [comment]
I made some progress today with the issues you may find on issue branch. So far I fixed packaging issue and the `UOcclusionPrimitiveContext::DebugBounds()` crash. I am still struggling a bit with the crash caused by `~TArray<FOcclusionMeshData,TSizedDefaultAllocator<32> >()`

Initially I assumed there is some race condition with the async code, but I seem to get the same behaviour with synchronous code as well. I am still looking into it.

**Comment by ArmainAP** [comment]
I merged all the fixed to main along d83361a4012410f526e8aa7124fb6d5ae769d19a
I am closing this issue for now as it got a bit past just compile errors. 

Please feel free to open up more issues if you encounter any other problems!
